### PR TITLE
Camel-quarkus-datasonnet extension - resolves Issue #2078

### PR DESCRIPTION
**Pull request by javaduke**

### CODE REVIEW

**Comment by aldettinger** [comment]
@javaduke Great to see that an extension is almost there :)

I wonder if the build issue on CI would be linked to inverting the order of bom imports. That switch might have consequences, we'll need to find an answer here.

Apart from that, a curiosity, the PR embeds dslibs.jar in the system under test for "datasonnet functionaly testing" purpose. Is it really needed to embed this ? For instance, couldn't it just be a test scoped dependency ?

**Comment by javaduke** [comment]
I think I found a better way to avoid pulling the old scala library. I'm testing it now.
The situation with `dslibs.jar` is different. This is just a resource inside the jar and the test verifies if datasonnet is able to read a library stored in a jar. With JVM tests I can just add the jar as a configuration parameter for the surefire plugin. This doesn't work for native tests though. I tried adding it to the pom as a test scope, but the problem is that this jar is not really a maven artifact, it's not in any maven repo, and `systemPath` cannot be used with any scope other than `system`. So I'm not sure if there's any other option to include this in the classpath for native tests.

**Comment by aldettinger** [comment]
Hey, @javaduke it's really lovely to see contributions :)

So, concerning the bom import order. It's better to change only at datasonnet level. I'm under the impression that it could still generate some issues with respect to enforcers and quarkus-platform. We'll see.

Now, regarding the dslibs.jar, I see a bit more. At this stage, I would merely remove this part from testing. We try to avoid jar in sources as far as possible. Plus, I would assume most camel-quarkus users would define there libs in their own app ? And also, some tricks might be needed in native mode. Anyway, that could be another ticket.

With ther2.8 release underway, it's maybe a better idea to review this PR next week with the embedded jar removed. That would already be a good first step :) Keep up the good work :+1:

**Comment by javaduke** [comment]
Ah, I was hoping to squeeze the data sonnet into the 2.8.0 release, because our work really depends on it.
Anyway, I removed the named imports test (no big deal, it is tested in `camel-datasonnet` anyway).

File: extensions/datasonnet/deployment/src/main/java/org/apache/camel/quarkus/component/datasonnet/deployment/DatasonnetProcessor.java
Comment (aldettinger) [inline extensions/datasonnet/deployment/src/main/java/org/apache/camel/quarkus/component/datasonnet/deployment/DatasonnetProcessor.java:42]: On scala version upgrade tests should capture this, so we are fine here IMO.

File: extensions/datasonnet/pom.xml
Comment (aldettinger) [inline extensions/datasonnet/pom.xml:56]: @ppalaga @jamesnetherton This PR changes the quarkus and cq pom import orders. I think there will be issues, especially with the plaform. Do you agree that we need to find another way or it could be fine ?

File: extensions/datasonnet/runtime/pom.xml
Comment (aldettinger) [inline extensions/datasonnet/runtime/pom.xml:37]: We may need to update the versions at some point in time.

File: extensions/datasonnet/runtime/pom.xml
Comment (aldettinger) [inline extensions/datasonnet/runtime/pom.xml:27]: Also here, the main branch will move to 2.9.0-SNAPSHOT in a moment.

**Comment by javaduke** [comment]
Sorry, I'm confused, would you like me to make any changes here?

File: extensions/datasonnet/deployment/src/main/java/org/apache/camel/quarkus/component/datasonnet/deployment/DatasonnetProcessor.java
Comment (aldettinger) [reply extensions/datasonnet/deployment/src/main/java/org/apache/camel/quarkus/component/datasonnet/deployment/DatasonnetProcessor.java:42]: Sorry, I was thinking aloud what would happen in case of scala version upgrade :)
And I think, the tests would fail, then we detect that some dependencies are not indexed, and then we update those lines of code. If you agree with my understanding, then I think we should keep this code as is.

**Comment by aldettinger** [comment]
Hey @javaduke, there is another PR that will update the version to 2.9.0-SNAPHOT https://github.com/apache/camel-quarkus/pull/3696. So once, #3696 is merged, then we would need to update the versions here.

Generally speaking the PR is good. The only issue I could see is the bom re-odering, I've asked help/confirmation about this. Still people are very busy. Worst case, I should be able to investigate further later on.

**Comment by javaduke** [comment]
OK, np, I'll keep an eye on the PR

**Comment by javaduke** [comment]
I bumped versions to 2.9.0 and committed again.

File: extensions/datasonnet/runtime/pom.xml
Comment (javaduke) [reply extensions/datasonnet/runtime/pom.xml]: Done.

File: extensions/datasonnet/runtime/pom.xml
Comment (javaduke) [reply extensions/datasonnet/runtime/pom.xml]: Done.

File: extensions/datasonnet/deployment/src/main/java/org/apache/camel/quarkus/component/datasonnet/deployment/DatasonnetProcessor.java
Comment (aldettinger) [reply extensions/datasonnet/deployment/src/main/java/org/apache/camel/quarkus/component/datasonnet/deployment/DatasonnetProcessor.java:42]: Reading again, I think my understanding is correct here. So we can keep like this.

File: extensions/datasonnet/runtime/pom.xml
Comment (javaduke) [reply extensions/datasonnet/runtime/pom.xml]: Done.

File: extensions/datasonnet/deployment/src/main/java/org/apache/camel/quarkus/component/datasonnet/deployment/DatasonnetProcessor.java
Comment (ppalaga) [inline extensions/datasonnet/deployment/src/main/java/org/apache/camel/quarkus/component/datasonnet/deployment/DatasonnetProcessor.java:55]: `combinedIndexBuildItem`and `reflectiveClasses` seem to be redundant. If not, please add a comment explaining what's their purpose.

File: extensions/datasonnet/pom.xml
Comment (ppalaga) [reply extensions/datasonnet/pom.xml:56]: Yes, `quarkus-bom` should always be before `camel-quarkus-bom`. Please remove the whole dependencyManagement section here, because the imports are there in the parent `camel-quarkus-extensions`.

**Review by ppalaga — CHANGES_REQUESTED** [review]

File: extensions/datasonnet/deployment/src/main/java/org/apache/camel/quarkus/component/datasonnet/deployment/DatasonnetProcessor.java
Comment (javaduke) [reply extensions/datasonnet/deployment/src/main/java/org/apache/camel/quarkus/component/datasonnet/deployment/DatasonnetProcessor.java]: Done.

**Comment by javaduke** [comment]
All done.

File: extensions/datasonnet/runtime/pom.xml
Comment (aldettinger) [reply extensions/datasonnet/runtime/pom.xml]: Great. Many thanks.

**Comment by ppalaga** [comment]
Thanks @javaduke, let's wait for the CI.

**Comment by javaduke** [comment]
Hmm, I'm not sure why is it failing, complains about generated flattened poms...

**Comment by javaduke** [comment]
Any updates on this? Is there anything else that is needed to be fixed?

File: integration-tests/datasonnet/pom.xml
Comment (aldettinger) [inline integration-tests/datasonnet/pom.xml:44]: Why is this dependencyManagement section needed ? It may lead to issue.

**Comment by javaduke** [comment]
It fails with complains that there are uncommitted changes:
```
Run [[ -z $(git status --porcelain | grep -v antora.yml) ]] || { echo 'There are uncommitted changes'; git status; exit 1; }
[10](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:10)
There are uncommitted changes
[11](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:11)
HEAD detached at pull/3686/merge
[12](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:12)
Changes not staged for commit:
[13](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:13)
  (use "git add <file>..." to update what will be committed)
[14](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:14)
  (use "git restore <file>..." to discard changes in working directory)
[15](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:15)
	modified:   poms/bom/src/main/generated/flattened-reduced-pom.xml
[16](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:16)
	modified:   poms/bom/src/main/generated/flattened-reduced-verbose-pom.xml
[17](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:17)

[18](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:18)
no changes added to commit (use "git add" and/or "git commit -a")
```

I have the same issue with the CXF extension CI build, and I can't reproduce it locally, so I have no idea what is wrong.

**Comment by javaduke** [comment]
I wonder if generated poms should not be in the repository, since they are generated by maven, maybe deleting them fixes the problem?

File: pom.xml
Comment (ppalaga) [inline pom.xml:131]: Scala version 2.12 is currently dictated by Quarkus. We could override it in our BOM, but any application generated by code.quarkus.io would get the version from Quarkus BOM anyway. 
Hence we have only two options here: (a) ask Quarkus to upgrade to Scala 2.13 or (b) make datasonnet work with 2.12.
I see this as the main issue of this PR.

**Review by ppalaga — CHANGES_REQUESTED** [review]
Please see inline. We need to figure out how we proceed with the Scala version.

**Comment by ppalaga** [comment]
> I wonder if generated poms should not be in the repository, since they are generated by maven, maybe deleting them fixes the problem?

We track the generated poms in git to have some long term auditing. E.g. in this PR, the changes in those file clearly show that you attempt to override scala libs maintained by Quarkus.

File: pom.xml
Comment (javaduke) [reply pom.xml:131]: Sorry, I don't think it's possible to make Datasonnet work with 2.12...so I guess we're in the holding pattern until Quarkus upgrades to 2.13

Comment (ppalaga) [reply pom.xml:131]: OK, thanks for confirming that it cannot work with Scala 2.12. There is this quite old quest to upgrade in Quarkus: https://github.com/quarkusio/quarkus/issues/10025 I have added a comment there.

File: integration-tests/datasonnet/pom.xml
Comment (ppalaga) [reply integration-tests/datasonnet/pom.xml:107]: Comparing the JSON documents as strings would not work here?

Comment (javaduke) [reply integration-tests/datasonnet/pom.xml:107]: No, because order is not guaranteed, and formatting may be off.

File: integration-tests/datasonnet/pom.xml
Comment (aldettinger) [reply integration-tests/datasonnet/pom.xml:72]: Indeed, this exclusion might not be needed. Logically, it could be possible to check by running `mvn depedency:tree` with and without the exclusion. The same dependency:tree should be resolved in both cases.

File: pom.xml
Comment (ppalaga) [reply pom.xml:131]: Scala 2.13 in Quarkus is underway https://github.com/quarkusio/quarkus/pull/25106

File: integration-tests/datasonnet/pom.xml
Comment (javaduke) [reply integration-tests/datasonnet/pom.xml]: Yup, this works, thank you!

**Comment by ppalaga** [comment]
@javaduke if you could please `rebase` and `push -f` rather than merge main. It is much easier for keeping an overview what is the actual state of your topic branch. Also in this project we prefer single commit per logical change. Future readers of the history might not be interested in all intermediate changes. Thanks for your great work!

**Comment by ppalaga** [comment]
The Scala upgrade in https://github.com/quarkusio/quarkus/pull/25106 is ready to be merged. You might want to build locally against that PR, check that it works for you and give feedback on that PR if necessary.

**Comment by javaduke** [comment]
I built quarkus locally and changed the version in the main pom.xml to 999-SNAPSHOT, but for some reason the flattened poms still pull 2.8.0.Final, and the resulting scala version is 2.12...

**Comment by aldettinger** [comment]
@javaduke I don't know exactly what is the issue at hand. Still, the quarkus-main branch may come in handy as it points to [quarkus last SNAPSHOT](https://github.com/apache/camel-quarkus/blob/quarkus-main/pom.xml#L59). Maybe relocating to this branch would be a good path ?

**Comment by javaduke** [comment]
Yes, I just tried it, same result - in my dependency tree I see `io.quarkus:quarkus-core:jar:999-SNAPSHOT:compile` which is what I expected, but I also see ` org.scala-lang:scala-library:jar:2.12.13:compile`.

**Comment by aldettinger** [comment]
@javaduke We are progressing, that's great :+1:  What about squashing all your commits into a single and opening another PR against the quarkus-main branch ? With such a PR, I feel we would be 2 or 3 details far from merging.

**Comment by javaduke** [comment]
Closing this PR since the new one was opened per @aldettinger request: https://github.com/apache/camel-quarkus/pull/3757/

**Pull request closed without merging**
